### PR TITLE
Unitnormmap

### PR DESCRIPTION
**Pull request by r-owen**
This branch adds a new kind of map: UnitNormMap, which computes a unit vector and its norm relative to a specified center. Thus the mapping has one more output than input. The intended use case is to support fairly arbitrary radial transformations, e.g. to represent optical distortion. Perhaps this is worth adding this to AST, though I am by no means certain of that. 

Note: I am proposing to add it to AST, not PyAST; this pull request is merely intended as a place to publicly discuss the code.

Known issues:
- unitnormmap.c does not specify a Rate function, even though an analytic Jacobian exists; it is a bit of a pain to have to compute the norm and I worry about making a mistake. But if it's important, this could be added.
- this branch does not include FORTRAN code

### CODE REVIEW

File: ast/unitnormmap.h [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+*     but WITHOUT ANY WARRANTY; without even the implied warranty of
+*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+*     GNU Lesser General Public License for more details.
+*     
+*     You should have received a copy of the GNU Lesser General
+*     License along with this program.  If not, see
+*     <http://www.gnu.org/licenses/>.
+
+*  Authors:
+*     DSB: D.S. Berry (Starlink)
```
Comment (timj) [inline ast/unitnormmap.h]: Author and history are not right.

**Comment by timj** [comment]
@dsberry would you mind taking a look?

**Comment by timj** [comment]
@r-owen I see that travis failed because AST hasn't quite decided what standard of C it's actually meant to be using and the Travis compiler is complaining about C99 usage.

File: ast/unitnormmap.h (hunk 1 above)
Comment (r-owen) [inline ast/unitnormmap.h]: Good catch. I will update this and the .c file. I will also check the other files I had to modify for history info that needs updating.

**Comment by r-owen** [comment]
It looks like the for loop variables may be the only issue. I'll change that.

**Comment by dsberry** [comment]
unitnormmap.c: 

1) The "Purpose" entry in the prologue needs changing
1b) In the "Description" entry, I'd also say how the output coordinate values are ordered - i.e. the lowest Nin output values are the unit vector and the highest output value (i.e. index Nin+1) is the norm.  Also include this in the public prologue for the constructor function.
2) The "Copyright" entry of the prologue needs changing, but I'm not sure what it should be @timj ? This also applies to the header file prologue.
3) A lot of the includes relating to other AST classes are no longer needed (matrixmap.h, winmap.h,  etc)
4) I'd use the astEQUAL macro defined in object.h instead of the local EQUAL macro. The local MAX, MIN and EQUAL macros could then be deleted (and this is probably the case in some  other classes too).
5) I'll just flag the fact that the convention within the rest of AST is to declare variables at the top of each function rather than where they are used. 
6) An extra step of sophistication for the MapMerge function would be to allow a combination of a ShiftMap followed by a forward UnitNormMap (or an inverted UnitNormMap followed by a ShiftMap) to be replaced by a single UnitNormMap with a different centre.  Also, a forward UnitNormMap followed by an inverted UnitNormMap with a different centre can be merged to create a ShiftMap.
7) In astINitUnitNormMap, you report an error if the centre array includes any bad values. I'm not sure this is a good idea. Other Mapping class behave differently - their initialisers accept bad values as parameter values, but  the Transform method then checks for this and returns bad output axis values.

**Comment by r-owen** [comment]
I'll fix 1, 1B, 3 (once I figure out which .h files aren't needed) and 4.
I have no idea what to do about the copyright.
Regarding (6): good point. If you are considering adding this to AST it is probably worth improving this.
Regarding (7): I originally coded it the way you requested but changed it to warn users right away and avoid a test for every call to "tran". I'll restore it.

**Comment by r-owen** [comment]
The code for merging mappings wants to check that centers are equal. Your comments for the EQUAL macro are:
/* Macro to check for equality of floating point values. We cannot
compare bad values directory because of the danger of floating point
exceptions, so bad values are dealt with explicitly. */
Can you please explain the issue? I would expect to be able to simply test if floats equal each other, since you are not using nan. (I was also surprised to see that the macro has a resolution of 1e-5, which seems large for testing equality; is exact equality too specific for this test?).

**Comment by r-owen** [comment]
As an aside: allowing bad values in centers also complicates testing if the centers are equal.

**Comment by r-owen** [comment]
I am borrowing an existing error code: AST__BADSM. Should I create a new one or borrow a different one?

**Comment by dsberry** [comment]
Regarding the EQUAL macro, I can't at the moment think why the explicit checks for AST__BAD were included. I've just checked the repo, and they go back almost to the dawn of AST  20 years ago, when things may have been different.  We could experiment with taking them out, but we'd need to be careful (also the use of a local EQUAL macro in other classes needs removing). 

Regarding the resolution of 1e-5, which is used , these values are to some extent arbitrary, but have been tweaked from time to time in the light of actual experience to solve specific issues faced by users.  I agree it's probably larger than necessary for your use in this mapping.  I suppose a new astEQUAL macro that allows a resolution to be specified as an argument would be a good idea. Until then, may be retain the local EQUAL macro with some suitably smaller resolution.  I don't think exact equality is a good idea because numerical mapping properties can acquire rounding errors as a consequence of mathematical operations when merging mappings. That is, if two Mappings are merged to create a single new Mapping, the numerical properties of the merged Mapping will  in general suffer from some rounding error. 

Regarding the AST__BADSM error code, I can't for the life of me think now why I included a check on map->shift in the shiftmap.c file. There's no way I can see a valid ShiftMap could be created with a NULL shift property, so I'm not sure why it is necessary to check for that in the Transform method. In general, AST does not test every pointer before using it - only those that could conceivably be NULL. So I'd just take out that check.

**Comment by r-owen** [comment]
I overhauled the simplification code, including supporting ShiftMap + UnitNormMap = UnitNormMap with new center. Unfortunately some of the work was wasted because ShiftMap is turned into WinMap before I get to it. Also I expanded a test that is showing a segfault in CmpMaps made with UnitNormMaps when I call tran on them. I have yet to track that down (made more difficult because because "python setup.py install" builds without debug symbols) but left the breakage in the unit tests on purpose.

**Comment by dsberry** [comment]
I admit I forgot about the way that ShiftMap is simplified, but I'd not say that your code is wasted as it is unwise to build in assumptions about the way that other classes get simplified (e.g. ShiftMap simplification could potentially be changed in the future for some reason). What it highlights is the possibility for merging a UnitNormMap with a WinMap, so long as the WinMap has unit scaling. I had overlooked that possibility.  

Regarding the segfault, it's usually much easier to track segfaults down in a regular AST application (written in C)  than it is in a pyast application.

**Comment by r-owen** [comment]
I expanded the simplification code to handle WinMap.

I found an easy way to enable debug symbols when building python extensions. My first attempt to track this down suggests the improbable answer of a bug in cmpmap's Transform method: at line 3779 it creates an intermediate point set based on the number of outputs of the compound mapping, rather than the first of the two mappings.

    temp2 = astPointSet( np, nout, "", status );

This fails with UnitNormMap because it produces more outputs than inputs. I am surly missing something obvious, since many other mappings have Nin!=Nout (though I don't know if any others have Nout > Nin in the forward direction).

**Comment by dsberry** [comment]
The intermediate PointSet used to hold the outputs from the first Mapping is "temp", not "temp2". It is created by the astTransform call at line  3790, which applies the first Mappings.

I think the segfault is caused by line 859 in unitnormmap.c:

int ncoord_out = astGetNcoord(out);

The "out" argument is supplied as NULL, which is what is causing the segfault. You should use "result" in place of "out". "result" is a copy of "out" if "out" is not NULL, or is a pointer to a new PointSet if "out" is NULL.

I also spotted a problem with threading in the astInitUnitNormMapVtab_  function.  In removing the "Local Variables:" block from the start of the function, you have also removed the declaration of the pointer to the structure holding thread-specific data. You need this line:

   astDECLARE_GLOBALS            /* Pointer to thread-specific global data */

at the start of each and every function that accesses any item of thread-specific data.

**Comment by r-owen** [comment]
Thank you, @dsberry. That was a big help. In testing my fix I found and fixed another error in how MapMerge set the invert flag when replacing a pair of maps with a simplified map (if the new map was inverted, that information was lost).

**Comment by dsberry** [comment]
So when it's ready we should add it to AST, which currently means cloning the whole Starlink repo. Tim is working on changing that, but in the meantime  if you don't feel like doing it yourself, I can do it for you.

**Comment by r-owen** [comment]
I think it's ready, though I am concerned about the unit tests. UnitNormMap is well tested in PyAST but the tests are written in python and so cannot be run in AST. I'm not sure how to proceed. I don't have a lot more time to work on this mapping.

**Comment by dsberry** [comment]
It would be a case of translating the tests into C or Fortran. Also I suppose there is still the Fortran wrapper file (funitnormmap.c) to write. Plus the docs. If you've run out of time I can take over if you like - it's up to you.

**Comment by r-owen** [comment]
If you feel the C code is in good shape then I would like to turn it over to you.

**Comment by timj** [comment]
I'm sorry I've not done the AST git extraction yet. It probably won't take long once I start it. I won't be able to do it today as I'm booked up with meetings but I can try and do it Wednesday. @dsberry does that work for you? (i.e. don't commit to AST in Starlink repo tomorrow).

**Comment by dsberry** [comment]
OK. Let me know when you've done it. Whilst I have your ear, what should the copyright statement be for the code that Russell has written. Not CCLRC presumably ....

**Comment by timj** [comment]
University of Washington

**Comment by r-owen** [comment]
Once AST is in its own repository, is there a plan to combine the official AST and PyAST into a single package? This would have many advantages to users:

- Python is widely used in Astronomy
- In enables writing AST unit tests in Python, making them easier to write. They also test two things at once: the Python wrappers and the underlying C code.
- LSST wants both the C and the Python wrappers

I don't know how the build system would work. Users who want C or FORTRAN access users need access to the libraries and header files. I hope it will be possible to build without FORTRAN (e.g. for Mac users, since the standard compiler does not support FORTRAN).

Failing that, could we at least make PyAST usable with the official AST for those users that require both the C code and Python code?

**Comment by timj** [comment]
My default thought would be to attach the AST git repo as a submodule to pyast as I do for palpy with PAL.

**Comment by dsberry** [comment]
There's a recently added --without-fortran configure option in AST.

**Comment by timj** [comment]
Next step for this pull request is to resync AST with the upstream repo that has the new code (Starlink/ast#1), and then rebase this PR on top of it. I am wondering whether we should turn `cextern/ast` into a submodule pointing directly to the standalone AST repo.

**Pull request closed without merging**